unittests/runtime/Mutex.cpp

```diff
@@ -362,3 +429,442 @@ TEST(MutexTest, ConditionLockOrWaitLockAndNotifyThreaded) {
+}
+
+TEST(ReadWriteLockTest, ReadLockCriticalSection) {
+  ReadWriteLock lock(/* reportFatal =*/ true, /* checked = */ true);
+  
+  int count = 0;
+  lock.readLock([&] { count++; });
+
+  ASSERT_EQ(count, 1);
+}
+
+TEST(ReadWriteLockTest, WriteLock) {
+  ReadWriteLock lock(/* reportFatal =*/ true, /* checked = */ true);
+
+  int count = 0;
+  lock.writeLock();
+  count++;
+  lock.unlock();
+
+  ASSERT_EQ(count, 1);
+}
+
+TEST(ReadWriteLockTest, TryWriteLock) {
+  ReadWriteLock lock(/* reportFatal =*/ true, /* checked = */ true);
+
+  int count = 0;
+  ASSERT_TRUE(lock.try_writeLock());
+  count++;
+  lock.unlock();
+
+  ASSERT_EQ(count, 1);
+}
+
+TEST(ReadWriteLockTest, ScopedWriteLock) {
+  ReadWriteLock lock(/* reportFatal =*/ true, /* checked = */ true);
+
+  int count = 0;
+  {
+    ScopedWriteLock guard(lock);
+    count++;
+  }
+
+  ASSERT_EQ(count, 1);
+}
+
+TEST(ReadWriteLockTest, ScopedWriteUnlock) {
+  ReadWriteLock lock(/* reportFatal =*/ true, /* checked = */ true);
+
+  lock.writeLock();
+
+  int count = 0;
+  {
+    ScopedWriteUnlock guard(lock);
+    count++;
+  }
+
+  lock.unlock();
+
+  ASSERT_EQ(count, 1);
+}
+
+TEST(ReadWriteLockTest, WriteLockCriticalSection) {
+  ReadWriteLock lock(/* reportFatal =*/ true, /* checked = */ true);
+
+  int count = 0;
+  lock.writeLock([&] { count++; });
+
+  ASSERT_EQ(count, 1);
+}
+
+TEST(ReadWriteLockTest, ReadWriteLockCriticalSection) {
+  ReadWriteLock lock(/* reportFatal =*/ true, /* checked = */ true);
+
+  int count = 0;
+  lock.readWriteLock(
+      [&] {
+        count++;
+        return false;
+      },
+      [&] { count++; });
+
+  ASSERT_EQ(count, 3);
+
+  count = 0;
+  lock.readWriteLock(
+      [&] {
+        count++;
+        return true;
+      },
+      [&] { count++; });
+
+  ASSERT_EQ(count, 1);
+}
+
+TEST(ReadWriteLockTest, ReadLockWhileReadLockedThreaded) {
+  ReadWriteLock lock(/* reportFatal = */ true, /* checked = */ true);
+
+  lock.readLock();
+
+  std::vector<bool> results;
+  results.assign(10, false);
+
+  bool done = false;
+  threadedExecute(10,
+                  [&](int index) {
+                    while (!done) {
+                      lock.readLock([&] {
+                        results[index] = true;
+                        std::this_thread::sleep_for(
+                            std::chrono::milliseconds(5));
+                      });
+                      std::this_thread::sleep_for(std::chrono::milliseconds(1));
+                    }
+                  },
+                  [&] {
+                    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+                    done = true;
+                  });
+
+  lock.unlock();
+
+  for (auto result : results) {
+    ASSERT_TRUE(result);
+  }
+}
+
+TEST(ReadWriteLockTest, ReadLockWhileWriteLockedThreaded) {
+  ReadWriteLock lock(/* reportFatal = */ true, /* checked = */ true);
+
+  lock.writeLock();
+
+  std::vector<int> results;
+  results.assign(10, 0);
+
+  bool done = false;
```

Comment:
Should be an atomic bool.  Applies to the other tests using "bool done" as well.